File: server/apps/api/src/animedownloader_api/routes/animes.py
from typing import Annotated
from uuid import UUID

from animedownloader_anime import (
    Anime,
    AnimeCreateData,
    AnimeService,
    AnimeUpdateData,
    Episode,
    EpisodeCreateData,
)
from fastapi import APIRouter, Depends, Response, status

from animedownloader_api.dependencies import get_anime_service
from animedownloader_api.schemas import (
    AnimeCreate,
    AnimeResponse,
    AnimeUpdate,
    EpisodeCreate,
)

router = APIRouter(prefix="/api/animes", tags=["animes"])
AnimeServiceDependency = Annotated[AnimeService, Depends(get_anime_service)]


@router.get("", response_model=list[AnimeResponse])
async def list_animes(service: AnimeServiceDependency) -> list[Anime]:
    return await service.list_animes()


@router.post("", response_model=AnimeResponse, status_code=status.HTTP_201_CREATED)
async def create_anime(
    payload: AnimeCreate,
    service: AnimeServiceDependency,
) -> Anime:
    return await service.create_anime(_to_anime_create_data(payload))


@router.get("/{anime_id}", response_model=AnimeResponse)
async def get_anime(
    anime_id: UUID,
    service: AnimeServiceDependency,
) -> Anime:
    return await service.get_anime(anime_id)


@router.patch("/{anime_id}", response_model=AnimeResponse)
async def update_anime(
    anime_id: UUID,
    payload: AnimeUpdate,
    service: AnimeServiceDependency,
) -> Anime:
    return await service.update_anime(
        anime_id,
        AnimeUpdateData(
            title=payload.title,
            year=payload.year,
            season=payload.season,
            weekday=payload.weekday,
            air_time=payload.air_time,
            timezone=payload.timezone,
        ),
    )


@router.delete("/{anime_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_anime(
    anime_id: UUID,
    service: AnimeServiceDependency,
) -> Response:
    await service.delete_anime(anime_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/{anime_id}/episodes", response_model=list[EpisodeResponse])
async def list_episodes(
    anime_id: UUID,
    service: AnimeServiceDependency,
) -> list[Episode]:
    return await service.list_episodes(anime_id)


@router.post(
    "/{anime_id}/episodes",
    response_model=EpisodeResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_episode(
    anime_id: UUID,
    payload: EpisodeCreate,
    service: AnimeServiceDependency,
) -> Episode:
    return await service.create_episode(
        anime_id,
        _to_episode_create_data(payload),
    )


def _to_anime_create_data(payload: AnimeCreate) -> AnimeCreateData:
    return AnimeCreateData(
        title=payload.title,
        year=payload.year,
        season=payload.season,
        weekday=payload.weekday,
        air_time=payload.air_time,
        timezone=payload.timezone,
        episodes=tuple(
            _to_episode_create_data(episode)
            for episode in payload.episodes
        ),
    )


def _to_episode_create_data(payload: EpisodeCreate) -> EpisodeCreateData:
    return EpisodeCreateData(
        episode_number=payload.episode_number,
        title=payload.title,
        source=payload.source,
        source_id=payload.source_id,
        source_title=payload.source_title,
        source_url=str(payload.source_url) if payload.source_url is not None else None,
        torrent_url=str(payload.torrent_url),
        size=payload.size,
        seeders=payload.seeders,
        leechers=payload.leechers,
        downloads=payload.downloads,
        info_hash=payload.info_hash,
        download_status=payload.download_status,
        conversion_status=payload.conversion_status,
    )

File: web/apps/web/src/entities/anime/api/animes.ts
import { z } from "zod";
import {
  deleteJson,
  getJson,
  patchJson,
  postJson,
} from "../../../shared/api/client";
import type { Anime, CreateAnimeInput, Episode, EpisodeInput } from "../model/types";

const episodeSchema = z.object({
  id: z.uuid(),
  anime_id: z.uuid(),
  episode_number: z.number().int().positive(),
  title: z.string(),
  source: z.string(),
  source_id: z.string().nullable(),
  source_title: z.string().nullable(),
  source_url: z.string().url().nullable(),
  torrent_url: z.string().url(),
  size: z.string().nullable(),
  seeders: z.number().int().nonnegative().nullable(),
  leechers: z.number().int().nonnegative().nullable(),
  downloads: z.number().int().nonnegative().nullable(),
  info_hash: z.string().nullable(),
  download_status: z.enum([
    "not_started",
    "downloading",
    "completed",
    "failed",
  ]),
  conversion_status: z.enum([
    "not_started",
    "converting",
    "completed",
    "failed",
  ]),
  created_at: z.coerce.date(),
  updated_at: z.coerce.date(),
});

const animeSchema = z.object({
  id: z.uuid(),
  title: z.string(),
  titles: z.record(z.string(), z.string()).default({}),
  year: z.number().int(),
  season: z.enum(["winter", "spring", "summer", "fall"]),
  weekday: z.enum([
    "monday",
    "tuesday",
    "wednesday",
    "thursday",
    "friday",
    "saturday",
    "sunday",
  ]),
  air_time: z.string().nullable(),
  timezone: z.string(),
  created_at: z.coerce.date(),
  updated_at: z.coerce.date(),
  episodes: z.array(episodeSchema),
});

function formatZodIssues(issues: z.core.$ZodIssue[]): string {
  return issues
    .map((issue) => {
      const path = issue.path.map(String).join(".");
      return path ? path + ": " + issue.message : issue.message;
    })
    .join("; ");
}

export class AnimeResponseError extends Error {
  constructor(readonly issues: z.core.$ZodIssue[]) {
    super(
      "Invalid anime API response: " + formatZodIssues(issues),
    );
    this.name = "AnimeResponseError";
  }
}

export async function listAnimes(signal?: AbortSignal): Promise<Anime[]> {
  const payload = await getJson("/api/animes", { signal });
  return parseList(payload);
}

export async function getAnime(id: string, signal?: AbortSignal): Promise<Anime> {
  return parseAnime(await getJson("/api/animes/" + id, { signal }));
}

export async function createAnime(
  input: CreateAnimeInput,
  signal?: AbortSignal,
): Promise<Anime> {
  return parseAnime(await postJson("/api/animes", input, { signal }));
}

export async function updateAnime(
  id: string,
  input: Partial<Omit<CreateAnimeInput, "episodes">>,
  signal?: AbortSignal,
): Promise<Anime> {
  return parseAnime(await patchJson("/api/animes/" + id, input, { signal }));
}

export async function deleteAnime(id: string, signal?: AbortSignal): Promise<void> {
  await deleteJson("/api/animes/" + id, { signal });
}

export async function createEpisode(
  animeId: string,
  input: EpisodeInput,
  signal?: AbortSignal,
): Promise<Episode> {
  return parseEpisode(
    await postJson("/api/animes/" + animeId + "/episodes", input, { signal }),
  );
}

export async function updateEpisode(
  id: string,
  input: Partial<EpisodeInput>,
  signal?: AbortSignal,
): Promise<Episode> {
  return parseEpisode(await patchJson("/api/episodes/" + id, input, { signal }));
}

export async function deleteEpisode(
  id: string,
  signal?: AbortSignal,
): Promise<void> {
  await deleteJson("/api/episodes/" + id, { signal });
}

function parseList(payload: unknown): Anime[] {
  const result = z.array(animeSchema).safeParse(payload);
  if (!result.success) {
    throw new AnimeResponseError(result.error.issues);
  }
  return result.data;
}

function parseAnime(payload: unknown): Anime {
  const result = animeSchema.safeParse(payload);
  if (!result.success) {
    throw new AnimeResponseError(result.error.issues);
  }
  return result.data;
}

function parseEpisode(payload: unknown): Episode {
  const result = episodeSchema.safeParse(payload);
  if (!result.success) {
    throw new AnimeResponseError(result.error.issues);
  }
  return result.data;
}
